### ISSUE TITLE
cumsum inaccuracy

### ISSUE THREAD

**Issue by clemisch** [google/jax#3517]
Dear jax team,

I am working on adding `np.unwrap` in jax. I have implemented the functionality by adapting the numpy source, but I'm struggling to get the tests working due to numerical inaccuracy for large arrays. 

I think the reason is `np.cumsum`. The error between numpy and jax grows roughly linearly with array size for both unwrap and cumsum. I have created a [colab notebook](https://colab.research.google.com/drive/19UcmQcKkbqTWQI4O3Ai2DZscFLFWp2lJ?usp=sharing) with example plots.

Is this inaccuracy expected with cumsum? If yes, how should I adapt tests for np.unwrap to account for this?

**Comment by jakevdp** [comment]
Thanks for the report! This is very curious... but digging into it, it appears that it might be numpy rather than jax that is the source of the inaccuracy:
```python
import numpy as np
import jax.numpy as jnp

npoints = 100_000
key = jax.random.PRNGKey(43)
x = jax.random.uniform(key, (npoints,)) * 100
x_np = np.asarray(x)

print("jnp.cumsum(x)[-1]:", jnp.cumsum(x)[-1])
print("np.cumsum(x)[-1]: ", np.cumsum(x_np)[-1])
print("np.sum(x):        ", np.sum(x_np))
print("jnp.sum(x):       ", jnp.sum(x))
```
The output is this:
```
jnp.cumsum(x)[-1]: 4982810.5
np.cumsum(x)[-1]:  4982824.0
np.sum(x):         4982810.0
jnp.sum(x):        4982810.5
```
Notice that the last entry of ``np.cumsum`` does not match numpy's sum of the values, as you would expect if the results were consistent. On the other hand, jax returns consistent results here.

I don't have any great ideas as to what might be causing this, but I'll try to dig a bit more into it.

**Comment by jakevdp** [comment]
It might be due to floating point accuracy. Take a look at this:
```python
import numpy as np
np.random.seed(0)
x32 = 100 * np.random.rand(100_000).astype('float32')
x64 = x32.astype('float64')
print(np.cumsum(x32)[-1])
print(x32.sum())
print(np.cumsum(x64)[-1])
print(x64.sum())
```
It appears that numpy's 32-bit cumsum accumulates rounding error in float 32.

**Comment by hawkinsp** [comment]
I'd have expected JAX's implementation to have good precision properties because it does a parallel prefix sum (https://developer.nvidia.com/gpugems/gpugems3/part-vi-gpu-computing/chapter-39-parallel-prefix-sum-scan-cuda). I have no idea what NumPy does.

**Comment by clemisch** [comment]
Thank you @jakevdp and @hawkinsp for the insight!

It appears like numpy uses more accurate pairwise sum only for `np.sum` but not for `np.cumsum` as mentioned [here](https://github.com/numpy/numpy/issues/9329). Even though it's said that the pairwise sum can't be provided for cumsum, may it be that jax manages exactly that with your parallel prefix sum?

In any case, my main concern is actually not the inaccuracy itself but rather to get `np.unwrap` into jax with working tests. What is best practice for a situation like this, where accuracy between numpy and jax is so different?

**Comment by jakevdp** [comment]
The best option would probably be to adjust the `tol` argument to the `_CheckAgainstNumpy` function in the test, with a comment mentioning this issue.

**Comment by mattjj** [comment]
@clemisch Thanks for this question; looking forward to your PR! In the meantime, I think we figured out the discrepancy, and @jakevdp 's recommendation sounds right, so I think it makes sense to close this issue. (Let me know if I'm mistaken!)
